rclrs/src/context.rs

```diff
@@ -118,6 +127,47 @@ impl Context {
     }
 }
 
+/// Returns non-ROS arguments held by context.
+///
+/// This function should be called after rcl_init. `rcl_arguments` should be `global_arguments`
+/// field of initialized context. `args` is array of input arguments passed to node.
+fn get_non_ros_arguments(
+    rcl_arguments: *const rcl_arguments_t,
+    args: &[*const c_char],
+) -> Result<Vec<String>, RclrsError> {
+    // SAFETY: In case of error -1 will be returned and handled a line below.
+    let args_status = unsafe { rcl_arguments_get_count_unparsed(rcl_arguments) };
+    if args_status == -1 {
+        return Err(RclrsError::RclError {
+            code: InvalidArgument,
+            msg: None,
+        });
+    }
+    // SAFETY: All possible negative args_status values handled above.
+    let args_num = args_status as usize;
+    if args_num == 0 {
+        return Ok(Vec::new());
+    }
+    let mut out_ptr: *mut i32 = null_mut();
+    unsafe {
+        // SAFETY: No preconditions for this function.
+        let allocator = rcutils_get_default_allocator();
+        // SAFETY: Error will be returned in case of broken rcl_arguments, no precondition for the rest.
+        rcl_arguments_get_unparsed(rcl_arguments, allocator, &mut out_ptr).ok()?;
+    }
+    // SAFETY: out_ptr won't be allocated in case of error and this code won't be reached.
+    // Indices of non-ROS args in input args array
+    let indices: Vec<i32> = unsafe { Vec::from_raw_parts(out_ptr, args_num, args_num) };
```

Comment:
According to the official [documentation](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts), this function is highly unsafe, and I'm not sure that all the invariants are accounted for...

For example, `out_ptr` needs to have been previously allocated via `String` or `Vec`, whereas our `out_ptr` was populated via a C string array, which isn't quite the same thing.

Is it possible to do this via any other method?

Reply:
My bad, didn't even check if C pointer can be freely deallocated by `Vec`.

Another approach would be parsing through memory allocated to `out_ptr` and call `deallocate` on `allocator` like this:

```rs
let mut indices = Vec::with_capacity(args_num);

for i in [0..args_num] {
    indices.push_back(*(out_ptr + i));
}

allocator.deallocate(out_ptr, null());
```
Note that it is safe to pass `NULL` to deallocate, since it's not used by [rcl_allocator_t implementation](https://github.com/ros2/rcutils/blob/master/src/allocator.c#L43).

The only safety precaution is that memory allocated during `rcl_arguments_get_unparsed` call has the exact same size as `rcl_arguments_get_count_unparsed` returns.